### ISSUE TITLE
Child template just does not render in a couple of scenarios after .NET 2.0 upgrade

### ISSUE THREAD

**Issue by rolandoldengarm** [toddams/RazorLight#134]
Just upgraded to .NET 2.0 beta 1.
It is working, compiling, etc. But in a couple of scenarios, my child HTML template just does not render. In the generated HTML there's just nothing there, also no errors.

E.g. this used to work in 1.1, not anymore in 2.0:
```
<tr style="@(rs.ActivityBookings == null || rs.ActivityBookings.Count == 0 ? "display:inherit" : "display:none")">
```
And when using a foreach, I now have to check if there are more than zero items. Without that check, again nothing rendered.

This is quite frustrating as this is trial and error, with no clue why this is not working. Am I doing something wrong?

**Comment by toddams** [comment]
1. What is child HTML template and how it is rendered? Example?
2. Did not get a thing about foreach. Example?

**Comment by rolandoldengarm** [comment]
Thanks for the reply @toddams 

1. In the main template I've got:
<div class="stops">
@{ await IncludeAsync("/Booking/StopContainer.cshtml", Model);  }
</div>
`StopContainer.cshtml` is quite long. But it was working in .NET Core 1.1, but stopped working in .NET 2.0. It would just render `<div class="stops"></div>`

As soon as I stripped almost everything from StopContainer.cshtml, it would render something again.

But if I'd add something like the <tr style ... > mentioned above, it would just render `<div class="stops"></div>` again

2. For each:

```
foreach (var hb in rs.HotelBookings)
{
 <h1>@hb.Title</h1>
}
```
HotelBookings is a collection. When HotelBookings has zero items, above HTML causes again the template to not render.
I had to change it to:
```
@if (rs.HotelBookings.Count > 0) 
{
foreach (var hb in rs.HotelBookings)
{
 <h1>@hb.Title</h1>
}
}
```
It is quite weird, but very annoying. The child template is quite long, and time consuming to analyze.
I've stopped at the point where rendering a certain property causes the template not to render.

By the way, no intellisense or debug warnings.

**Comment by rolandoldengarm** [comment]
@toddams when I compile the child template directly, it works fine.
It seems to have something to do with the rendering of child templates.

**Comment by atjensw** [comment]
I've also have trouble with child template. 
Snippet from main template:
```
 @foreach (var subscription in Model.Subscriptions)
            {
                <table class="outer" align="center" style="border-spacing: 0; font-family: arial, sans-serif; color: #333333; Margin: 0 auto; width: 100%; max-width: 600px;">
                   
                    <tr>
                        <td class="one-column" style="padding-top: 0; padding-bottom: 0; padding-right: 0; padding-left: 0;">

                            <!-- Properties start -->
                            @{  await IncludeAsync("Parts.SubscriptionMail.Property.cshtml", subscription); }
                            <!-- Properties end -->
                        </td>
                    </tr>

```
Child template looks like this:
```
@using RazorLight
@using MyNamespace.TemplateModels.Subscription;
@using MyNamespace.TemplateModels;
@using DMyNamespace.TemplatePages;
@inherits SubscriptionTemplatePage<SubscriptionMailModel>

<table style="border-spacing: 0; font-family: arial, sans-serif; color: #333333; width: 100%;">
    @foreach (var property in Model.Properties)
    {
        <tr>
            <td class="contents" style="text-align: center; background-color: #fff; line-height: 24px; width: 100%; padding-top: 0; padding-bottom: 0; padding-right: 0; padding-left: 0;">
                <a href="http://test.com/@property.Id" style="color: #007272; text-decoration: none;">
                    <img src="@property.MainImage" width="100%" height="auto" style="border-width: 0;" />
                </a>
                <p style="Margin: 0; font-size: 14px; Margin-bottom: 10px;">&nbsp;</p>

                <a class="size-24" href="http://test.com/@property.Id" style="color: #007272; text-decoration: none; font-size: 24px;">
                    @Helper.GetHeader(property)
                </a><br />

                @Helper.GetSubHeader(property)<br />

                <strong>BOLIGTYPE:</strong>
                <span class="gray" style="color: #666666;">
                    @Helper.GetObjectType(property)
                </span><br />

                @Raw(@Helper.GetPrice(property))
                <p style="Margin: 0; font-size: 14px; Margin-bottom: 10px;">&nbsp;</p>

                <!--[if mso]>
                    <v:roundrect xmlns:v="urn:schemas-microsoft-com:vml" xmlns:w="urn:schemas-microsoft-com:office:word" href="http://test.com/@property.Id" arcsize="10%" stroke="f" fillcolor="#f87f00" style="height: 40px; v-text-anchor: middle; width: 200px;">
                    <w:anchorlock/>
                    <center>
                <![endif]-->
                <a class="btn-show-property" href="http://test.com/@property.Id" style="text-decoration: none; display: block; background-color: #f87f00; color: #ffffff; margin-top: 0; margin-bottom: 0; margin-right: auto; margin-left: auto; padding-top: 10px; padding-bottom: 10px; padding-right: 0; padding-left: 0; border-radius: 4px; font-size: 13px; font-weight: bold; width: 200px; font-family: arial, sans-serif;">Vis boligen</a>
                <!--[if mso]>
                    </center>
                    </v:roundrect>
                <![endif]-->
                <p style="Margin: 0; font-size: 14px; Margin-bottom: 10px;">&nbsp;</p>

            </td>
        </tr>
        <tr>
            <td class="row-padder" style="padding-top: 0; padding-bottom: 0; padding-right: 0; padding-left: 0; background-color: inherit;">&nbsp;</td>
        </tr>
    }
    </table>
```
This works when i have less then 5 objects in the subscription.Properties property. If i have more then 5 then it will stop rendering after  ` <!--  Properties start -->`. I don't get any exceptions but i get a broken template.

**Comment by toddams** [comment]
I found the bug, and that is a total disaster!) 
Can't imagine how we've been using this for such a long time without facing this bug, but with long templates it happens

Will prepare the next release, thank you for the very detailed description

**Comment by sbinkert** [comment]
@toddams  When do you guess the nuget package will be released for this fix?

**Comment by yfayerman** [comment]
I also hit the same bug, child templates with a loop over a collection with the size more than 4 elements would not render at all.

@toddams: could you share details of the fix if you have one?

**Comment by favishG** [comment]
I also have this issue. After playing around i saw that the problem (by me) only happens if i have a foreach loop within another foreach loop in a child view.

Can you please provide details of the fix?

**Comment by AstralRoad** [comment]
Hi guys, i simply changed the file 

        src\RazorLight\Internal\Buffering\ViewBuffer.cs  line21
```
        public static readonly int ViewPageSize = 512; //(from 256)
```
my template works fine,   then change back to 256 , it's broken again.  
after set it to 512, works ; then provide a large mumber items  to generate, fail again,  

I 'm sure that  the **ViewPageSize** is the point.  sth generated bursted the "buffer".

so i temporary set it to a big enough number. 

@toddams @favishG

**Comment by toddams** [comment]
@AstralRoad 
There is an issue with flushing the buffer after each partial iteration. Increasing the ViewPageSize will just move the threshold, but issue could occur again on larger templates. I'm working on a fix

**Comment by matteggg** [comment]
Do you have a time plan for when this bugfix will be released?

**Comment by Ahoo-Wang** [comment]
@toddams 
I used [RazorLight - 2.0.0-beta1](https://www.nuget.org/packages/RazorLight/2.0.0-beta1) in my open source project [SmartCode](https://github.com/Ahoo-Wang/SmartCode). I also found such a serious bug.
When can you post a new version to nuge to fix this bug?
Thank you very much!!!

**Comment by wik8wz** [comment]
Commit 0dbc5e1d174411530c3ab81db94dfc3444dd4a4d does not fix rendering issue. It is still occurring for large templates.  The problem is caused by MemoryPoolViewBufferScope implementation (clearing/passing/returning the buffer).
I implemented workaround here https://github.com/wik8wz/RazorLight/commit/35f06da97ec7da7677379932397560410a44d04b

**Comment by markleacompassphs** [comment]
Any progress on this?  We have already invested a great deal of time on a large project and then ran into this issue once our template got to big.

**Comment by wik8wz** [comment]
We went live with the fork which I created. Might switch back to original one when fixed.

**Comment by ADNewsom09** [comment]
> We went live with the fork which I created. Might switch back to original one when fixed.

@wik8wz Can you open a PR from your fork with the fix?

**Comment by Lukxxx** [comment]
> @AstralRoad
> There is an issue with flushing the buffer after each partial iteration. Increasing the ViewPageSize will just move the threshold, but issue could occur again on larger templates. I'm working on a fix

@toddams could you please release a nuget beta package with that fix?

**Comment by hhgm** [comment]
I hit what sounds like this bug today.... 

Fairly simple main view:

@using Microsoft.EntityFrameworkCore.Internal
@using RazorLight
@using TMServerDb.Models.Core
@inherits RazorLight.TemplatePage<TMServerAPI.Models.Reporting.ReportViewModel>
@{
    DisableEncoding = true;
}
<!doctype html>
<html lang="en">
<head>
    <link rel="stylesheet" href="~/Bootstrap/css/bootstrap.css" />
    <title>@Model.Definition.Name</title>
</head>
<body>
    @{
        await IncludeAsync("..\\StandardItems\\TournamentHeader.cshtml", (Model.Data as TMServerDb.Models.Core.MatchSet).Tournament);
    }
    <br>
    @{
        var ms = Model.Data as TMServerDb.Models.Core.MatchSet;
        var grouped = ms.Matches.OrderBy(n => n.Number).ThenBy(p => p.Pitch.Name).GroupBy(s => s.PlannedStart);

        foreach (IGrouping<DateTime, Match> grouping in grouped)
        {
            await IncludeAsync("..\\StandardItems\\MatchGrouping.cshtml", grouping);

            <br/>
        }
    }
</body>
</html>

The nested view:

@using RazorLight
@using TMServerDb.Models.Core
@inherits RazorLight.TemplatePage<IGrouping<DateTime, TMServerDb.Models.Core.Match>>
<table class="table .table-striped .table-light">
    <thead>
    <tr>
        <th style="width: 100px">@Model.Key.Date.ToShortDateString()</th>
        <th style="alignment: center">@Model.Key.ToShortTimeString()</th>
        <th style="width: 100px"></th>
    </tr>
    <tr>
        <th scope="col">#</th>
        <th scope="col">Pitch</th>
        <th scope="col">Class</th>
    </tr>
    </thead>
    <tbody>
    @{
        foreach (Match match in Model.OrderBy(x => x.Pitch.Name))
        {
            <tr>
                <th scope="row">@match.Number</th>
                <td>@match.Pitch.Name</td>
                <td>@match.Class.Name</td>
            </tr>
        }
    }
    </tbody>
</table>

When rendering, in my test, I have 30 odd groupings... but only a few make it to the pdf... intercepting the result from the rendering, explains why... notice the numerous ending of html and body tags...

<!doctype html>
<html lang="en">
<head>
    <link rel="stylesheet" href="~/Bootstrap/css/bootstrap.css" />
    <title>Tournament Schedule</title>
</head>
<body>
<div>
    <h1 align="center">Amsterdam Open 2019</h1>
    <h2 align="center">Uithoorn</h2>
    <h3 align="center">01-06-2019 - 02-06-2019</h3>
</div>
    <br>
<table class="table .table-striped .table-light">
    <thead>
    <tr>
        <th style="width: 100px">01-06-2019</th>
        <th style="alignment: center">09:00</th>
        <th style="width: 100px"></th>
    </tr>
    <tr>
        <th scope="col">#</th>
        <th scope="col">Pitch</th>
        <th scope="col">Class</th>
    </tr>
    </thead>
    <tbody>
            <tr>
                <th scope="row">1</th>
                <td>Pitch 1</td>
                <td>4th Division</td>
            </tr>
            <tr>
                <th scope="row">2</th>
                <td>Pitch 2</td>
                <td>1st Division</td>
            </tr>
            <tr>
                <th scope="row">3</th>
                <td>Pitch 3</td>
                <td>Women Division</td>
            </tr>
            <tr>
                <th scope="row">4</th>
                <td>Pitch 4</td>
                <td>Women Division</td>
            </tr>
            <tr>
                <th scope="row">5</th>
                <td>Pitch 5</td>
                <td>2nd Division</td>
            </tr>
            <tr>
                <th scope="row">6</th>
                <td>Pitch 6</td>
                <td>3th Division</td>
            </tr>
    </tbody>
</table>            <br/>
<table class="table .table-striped .table-light">
    <thead>
    <tr>
        <th style="width: 100px">01-06-2019</th>
        <th style="alignment: center">09:30</th>
        <th style="width: 100px"></th>
    </tr>
    <tr>
        <th scope="col">#</th>
        <th scope="col">Pitch</th>
        <th scope="col">Class</th>
    </tr>
    </thead>
    <tbody>
            <tr>
                <th scope="row">7</th>
                <td>Pitch 1</td>
                <td>4th Division</td>
            </tr>
            <tr>
                <th scope="row">8</th>
                <td>Pitch 2</td>
                <td>2nd Division</td>
            </tr>
            <tr>
                <th scope="row">9</th>
                <td>Pitch 3</td>
                <td>1st Division</td>
            </tr>
            <tr>
                <th scope="row">10</th>
                <td>Pitch 4</td>
                <td>2nd Division</td>
            </tr>
            <tr>
                <th scope="row">11</th>
                <td>Pitch 5</td>
                <td>1st Division</td>
            </tr>
            <tr>
                <th scope="row">12</th>
                <td>Pitch 6</td>
                <td>3th Division</td>
            </tr>
    </tbody>
</table>            <br/>
<table class="table .table-striped .table-light">
    <thead>
    <tr>
        <th style="width: 100px">01-06-2019</th>
        <th style="alignment: center">10:00</th>
        <th style="width: 100px"></th>
    </tr>
    <tr>
        <th scope="col">#</th>
        <th scope="col">Pitch</th>
        <th scope="col">Class</th>
    </tr>
    </thead>
    <tbody>
            <tr>
                <th scope="row">13</th>
                <td>Pitch 1</td>
                <td>4th Division</td>
            </tr>
            <tr>
                <th scope="row">14</th>
                <td>Pitch 2</td>
                <td>Women Division</td>
            </tr>
            <tr>
                <th scope="row">15</th>
                <td>Pitch 3</td>
                <td>2nd Division</td>
            </tr>
            <tr>
                <th scope="row">16</th>
                <td>Pitch 4</td>
                <td>2nd Division</td>
            </tr>
            <tr>
                <th scope="row">17</th>
                <td>Pitch 5</td>
                <td>1st Division</td>
            </tr>
            <tr>
                <th scope="row">18</th>
                <td>Pitch 6</td>
                <td>3th Division</td>
            </tr>
    </tbody>
</table>            <br/>
<table class="table .table-striped .table-light">
    <thead>
    <tr>
        <th style="width: 100px">01-06-2019</th>
        <th style="alignment: center">10:30</th>
        <th style="width: 100px"></th>
    </tr>
    <tr>
        <th scope="col">#</th>
        <th scope="col">Pitch</th>
        <th scope="col">Class</th>
    </tr>
    </thead>
    <tbody>
            <tr>
                <th scope="row">19</th>
                <td>Pitch 1</td>
                <td>4th Division</td>
            </tr>
            <tr>
                <th scope="row">20</th>
                <td>Pitch 2</td>
                <td>1st Division</td>
            </tr>
            <tr>
                <th scope="row">21</th>
                <td>Pitch 3</td>
                <td>Women Division</td>
            </tr>
            <tr>
                <th scope="row">22</th>
                <td>Pitch 4</td>
                <td>1st Division</td>
            </tr>
            <tr>
                <th scope="row">23</th>
                <td>Pitch 5</td>
                <td>2nd Division</td>
            </tr>
            <tr>
                <th scope="row">24</th>
                <td>Pitch 6</td>
                <td>3th Division</td>
            </tr>
    </tbody>
</table>            <br/>
            <br/>
</body>
</html>            <br/>
</body>
</html>            <br/>
</body>
</html>            <br/>
</body>
</html>            <br/>
</body>
</html>            <br/>
</body>
</html>            <br/>
</body>
</html>

**Comment by MNF** [comment]
@wik8wz , is your NoPoolingViewBufferScope work fine in production? Have you noticed any issues like high memory consumption or memory leak?

**Comment by wik8wz** [comment]
> @wik8wz , is your NoPoolingViewBufferScope work fine in production? Have you noticed any issues like high memory consumption or memory leak?

Nothing was reported until now. We do recycle our app pool everyday anyway. 
Our service is not very loaded at the moment. We are sending ~250 rendered email per month right now

**Comment by markleacompassphs** [comment]
Big thanks to wik8wz for fixing this issue.  We are using your branch and no longer have the partial view issues that we were facing.  Another interesting thing we found was that you need to be careful about making var assignments inside of the partial views, you are better served to just directly looking into your model for all of your data.

**Comment by MNF** [comment]
@toddams , are you ok with https://github.com/wik8wz/RazorLight/commit/35f06da97ec7da7677379932397560410a44d04b implementation? Will you accept PR with it? We are also using his branch in a couple of projects without any issues.

**Comment by wik8wz** [comment]
@markleacompassphs in meantime we usage of service in production increased significantly to 1000s emails per month. Did not face rendering issue again since fix was implemented (over 8 months now). Quite happy how it is working to be honest.

**Comment by jzabroski** [comment]
I recently requested write access to this repo form @toddams  in order to help out with the project and get a few key bugs fixed.

@wik8wz  Do you have any interest in a pull request for your fix?  I'd rather you formally submit one than me copy-paste your contribution, on legal grounds.

**Comment by wik8wz** [comment]
@jzabroski I created one here: https://github.com/toddams/RazorLight/pull/275 Sadly I'm no longer working with C# and therefore will no longer support development here. I hope this works for you (I saw there are no merge conflicts)

**Comment by jzabroski** [comment]
@wik8wz  Why sadly?  Friend request me on LinkedIn, who knows, maybe if you love C# I would love to work with you.

Thanks for your help!

**Comment by BuairtRi** [comment]
Adding my note here that @wik8wz 's fork seems to have fixed the problem for me. Much thanks!

@toddams is there any chance we can get this fix integrated? Or at least add a disclaimer to the README on the repo? I spent the better part of the day tracking down a solution. Was afraid I'd have to switch to an entirely different Razor rendering library.

**Comment by jzabroski** [comment]
I'm waiting on @toddams to give me access to update the Nuget package, then within a few days I should be able to fix all these problems.  He gave me write access to the repo, but not the nuget feed.

**Comment by markleacompassphs** [comment]
@jzabroski  That's great news - we are still having great success with this branch, in fact we are getting ready to deploy our new communication service (replacing a horribly done WCF from years back) with the new .net core/Razor Light that has a pretty slick data augmentation architecture.  Keep us posted!

**Comment by jzabroski** [comment]
Duplicate: #261

**Comment by jzabroski** [comment]
@markleacompassphs Can you please confirm if the latest 2.0.0-beta2 release fixes this problem?  I feel pretty confident it does.

**Comment by jzabroski** [comment]
@markleacompassphs Just a nudge - I believe this is fixed.  Looking for someone else to confirm.

**Comment by markleacompassphs** [comment]
@jzabroski we are not having any issues with the partial views with big templates ever since we switched to using the branch that @wik8wz created a long time ago.  We have just been using his source code in our solution we have it as a project and we can debug as needed.  We did have an issue with the cache yesterday, but so far after a rebuild and re-deploy it appears to be working fine.  We are scheduled to go into Production this weekend.
